Explorer/Assets/Scripts/SceneRuntime/Apis/Modules/EngineApi/SDKObservableEvents/SDKObservableEventsEngineApiWrapper.cs

```diff
@@ -0,0 +1,257 @@
+using System.Collections.Generic;
+
+namespace SceneRuntime.Apis.Modules.EngineApi.SDKObservableEvents
+{
+    public class SDKObservableEventsEngineApiWrapper : EngineApiWrapper
+    {
+        private static List<SDKObservableEvent> EMPTY_EVENTS_LIST = new ();
+
+        private readonly PBAvatarEmoteCommand avatarEmoteCommand = new ();
+        private readonly ProtobufSerializer<PBAvatarEmoteCommand> avatarEmoteCommandSerializer = new ();
+        private readonly PBPlayerIdentityData playerIdentityData = new ();
+        private readonly ProtobufSerializer<PBPlayerIdentityData> playerIdentityDataSerializer = new ();
+        private readonly PBRealmInfo realmInfo = new ();
+        private readonly ProtobufSerializer<PBRealmInfo> realmInfoSerializer = new ();
+        private readonly List<SDKObservableEvent> sdkObservableEvents = new ();
+        private readonly HashSet<string> sdkObservableEventSubscriptions = new ();
+        private readonly Dictionary<CRDTEntity, string> userIdEntitiesMap = new ();
+        private readonly ISDKObservableEventsEngineApi engineApi;
+        private readonly ISDKMessageBusCommsControllerAPI commsApi;
+        private bool reportedSceneReady;
+
+        public SDKObservableEventsEngineApiWrapper(ISDKObservableEventsEngineApi api, ISDKMessageBusCommsControllerAPI commsApi, IInstancePoolsProvider instancePoolsProvider, ISceneExceptionsHandler exceptionsHandler) : base(api, instancePoolsProvider, exceptionsHandler)
+        {
+            this.engineApi = api;
+            this.commsApi = commsApi;
+        }
+
+        public override void Dispose()
+        {
+            sdkObservableEvents.Clear();
+            sdkObservableEventSubscriptions.Clear();
+            base.Dispose();
+        }
+
+        // Used for SDK Observables + SDK Comms MessageBus
+        [UsedImplicitly]
+        public List<SDKObservableEvent> SendBatch()
+        {
+            if (sdkObservableEventSubscriptions.Count == 0)
+            {
+                engineApi.ClearOutgoingCRDTMessages();
+                return EMPTY_EVENTS_LIST;
+            }
+
+            try
+            {
+                sdkObservableEvents.Clear();
+
+                DetectObservableEventsFromComponents();
+
+                // SDK 'comms' observable for scenes MessageBus
+                DetectSceneMessageBusCommsObservableEvent();
+
+                return sdkObservableEvents;
+            }
+            catch (Exception e)
+            {
+                // Report an uncategorized MANAGED exception (don't propagate it further)
+                exceptionsHandler.OnEngineException(e);
+                return EMPTY_EVENTS_LIST;
+            }
+        }
+
+        private void DetectObservableEventsFromComponents()
+        {
+            // 1. Due to uncertainty in CRDT messages order we have to first check those which populate the userIdEntitiesMap
+            foreach (ProcessedCRDTMessage outgoingCRDTMessage in engineApi.OutgoingCRDTMessages)
```

Comment:
You can create a map in `SDKObservableEventsEngineAPIImplementation` straight-away as you process individual messages, you don't need to iterate over all of them once more

Reply:
True, ended up creating a separate `priorityOutgoingCRDTMessages` that only contain the playerIdentityData `PUT_COMPONENT` messates, [here](https://github.com/decentraland/unity-explorer/pull/662/commits/941835f64eb8c016a10744dfd6a21aab6d7b5b9e)

Reply:
Ended up refactoring a bit and moving the observable-replacement components detection logic directly into the `SDKObservableEventsEngineAPIImplementation` class, **to detect messages right away and avoid traversing them all again in the wrapper class**.
Here: https://github.com/decentraland/unity-explorer/pull/662/commits/6a06242775a4a75c5ce3acb424cf49427c56c8d1